[CALCITE-3913] Test correctness using formal verification (Qi Zhou)

qizhou92 opened the pull request:
Add dependency for z3, add two simple tests to test if native library work.
I am new to this process, and I am not quite sure how much code I should push in commit to reviewing.
Welcome any feedback or propose, given the relatively large number of codes that I need to implement this function.
The instruction for building a native library for z3 can be found here: https://github.com/Z3Prover/z3. This instruction should be easy to follow. Please using the z3-4.41 version to build the native library.
Add symbolic columns, and convert simple rex node to symbolic columns

qizhou92 committed:
[CALCITE-3913] Test correctness using formal verification (Qi Zhou). Add dependency for z3, add two simple test to test if native library work.

fix error for smtLibTest

qizhou92 committed:
[CALCITE-3913]Test Correctness using formal verifcation (Qi Zhou), add symbolic columns

qizhou92 committed:
[CALCITE-3913]Test Correctness using formal verification(Qi Zhou), Rex2SymbolicColumn

[CALCITE-3913]Test correctness using formal verification(Qi Zhou)

fix error for rextosymbolicCOlumn

danny0405 commented:
I guess maybe  @vlsi can give a good review ~

vlsi commented on core/build.gradle.kts:
```diff
@@ -60,6 +60,7 @@ dependencies {
     implementation("org.codehaus.janino:commons-compiler")
     implementation("org.codehaus.janino:janino")
     implementation("org.slf4j:slf4j-api")
+    implementation("org.sosy-lab:javasmt-solver-z3")
```
```suggestion
    testImplementation("org.sosy-lab:javasmt-solver-z3")
```

vlsi commented on core/src/test/java/org/apache/calcite/test/verifier/SymbolicColumn.java:
```diff
@@ -0,0 +1,91 @@
+
+public class SymbolicColumn {
+  /** Each SymbolicColumn is unique. **/
+  private static int count = 0;
```
It would be better to use `AtomicInteger` just in case.

vlsi commented:
@qizhou92 , thanks for the PR.

>  I am not quite sure how much code I should push in commit to reviewing

That differs on a case-by-case basis.

For z3 integration, I would suggest having at least one Calcite-specific test implemented. In other words, currently, all the added z3 tests verify only that z3 loads.

It would be nice if you could add a test that verifies if the two RexNodes are equivalent or not (or whatever was your intention with z3).

vlsi commented on core/src/test/java/org/apache/calcite/test/verifier/SmtLibTest.java:
```diff
@@ -0,0 +1,72 @@
+      solver.add(eq);
+      assertEquals(solver.check(), Status.SATISFIABLE);
+    } catch (Error e) {
+      System.err.println("cannot load native library libz3java ");
```
Please move "z3 availability" check to `@BeforeAll` method, so all the tests are marked as skipped in case the library is not available.

The test could `throw Exception`, so you don't need to catch exceptions.

qizhou92 committed:
[CALCITE-3913] Test correctness using formal verification (Qi Zhou), add test cases

qizhou92 commented:
Thank you @vlsi  for your review. I update the code based on your comments. I move the Smtlib.test file out of the verifier folder so that I can reuse some of the testing infra there. I create two test cases to checking if two RexNodes are equivalent. Do you think it is the right path?

vlsi commented on core/src/test/java/org/apache/calcite/test/SmtLibTest.java:
```diff
@@ -0,0 +1,115 @@
+    Context z3Context = new Context();
+    List<SymbolicColumn> inputSymbolicColumns = new ArrayList<>();
+    /** mock emp table, 8 columns with int type, since we only support numerical type for now **/
+    for (int i = 0; i < 8; i++) {
+      RelDataType intType = typeFactory.createJavaType(int.class);
+      SymbolicColumn inputColumn = SymbolicColumn.mkNewSymbolicColumn(z3Context, intType);
+      inputSymbolicColumns.add(inputColumn);
```
Please use `RexProgramTestBase` so you could reuse all the helpers for types, etc, etc.

qizhou92 replied:
Hi, thank you for your comments. I use **RexProgramBuilderBase**, which solves the problem. And **RexProgramTestBase** is not a public class.

vlsi replied:
Would you please move your classes to `org.apache.calcite.rex` package? Then `RexProgramTestBase` would be visible.

qizhou92 replied:
Sorry, I am a little bit confused about how to skip test cases if the z3 dynamic library is not loaded by using @BeforeAll. It seems even I throw errors in the BeforeAll methods, it still fail the test cases, which break the whole CI.

vlsi replied:
> It seems even I throw errors in the BeforeAll methods, it still fail the test cases, which break the whole CI.

You need to catch the exception and call `Assumptions.assume...`
If the test fails with `assumption violated`, then it would be marked as skipped.

See https://junit.org/junit5/docs/5.0.0/api/org/junit/jupiter/api/Assumptions.html#assumeTrue-boolean-

For instance:

```java
Assumptions.assumeTrue(verifyZ3Loads(), "z3 is required for the tests")
```

vlsi replied:
Right you are. The proper base class is `RexProgramBuilderBase`

qizhou92 committed:
[CALCITE-3913]Test correctness using formal verification (Qi Zhou), fix test cases

qizhou92 committed:
[CALCITE-3913] Test correctness uisng formal verification (Qi Zhou)

[CALCITE-3913] Test correctness using formal verification (Qi Zhou), fix test cases

qizhou92 committed:
[CALCITE-3913] Test correctness uisng formal verification (Qi Zhou)

vlsi commented on core/src/test/java/org/apache/calcite/test/SmtLibTest.java:
```diff
@@ -0,0 +1,109 @@
+  }
+
+  @Test void rexNodeEq1() {
+    final String cond1 = " empno > 10 and deptno = 5";
+    final String cond2 = " empno + deptno > 15 and deptno = 5";
```
This is nice. However, it might result in accidental "automatic simplification" during parsing.
In other words, parser (or SQL to rex converter) might silently simplify the expression, so the actual tested node would be different from what is represented in the sources.

On the other hand, `RexProgramBuilderBase` provides methods like `and`, `or`, etc. so `RexNode` can be created with fine details.

vlsi commented:
@qizhou92 , this looks nice.
Have you seen `RexFuzzer`, `RexProgramFuzzyTest`, `org.apache.calcite.rex.RexSimplify#verify` ?

I wonder if `z3` verifications could be integrated to `RexSimplify#verify`. Of course, it would be limited to `int` or something, but it still might be valuable.

An alternative option is to add a dedicated case to `RexProgramFuzzyTest` which would generate `int-only` randomized expressions (e.g. via `RexFuzzer`), and verify if `RexSimplfy` does not corrupt them.

qizhou92 commented:
@vlsi , Yes, it can be used for RexSimplify#verify as the first use cases. I will add more codes to support SPJ RelNode, aggregate node, union node, outer join node, step by step.

qizhou92 committed:
[CALCITE-3913] Test correctness using formal verification (Qi Zhou), fix test cases

vlsi commented on core/src/test/java/org/apache/calcite/test/SmtLibTest.java:
```diff
@@ -0,0 +1,116 @@
+      solver.check();
+      z3Loads = true;
+    } catch (Error e) {
+      z3Loads = false;
```
Does `assume` work here, so it is not duplicated across all the tests?

qizhou92 committed:
[CALCITE - 3913] Test correctness using formal verification (Qi Zhou)

[CALCITE-3913] Test correctness using formal verification (Qi Zhou)

qizhou92 commented:
Hi, I implement the function that verifies the equivalence of two rex node, including numerical value, and a boolean value. I use some test cases previously in RexProgromTest to verify the correctness of RexSimplify. I can keep working support more constructors such as CASE, or more types, such as DATE. 
another question I have is I found there is a class is called RexImplicationChecker, Z3 can do a better job to check the implication between two RexNode. For example, the current RexImplicationChecker seems cannot verify that Not (a <=30 ) ==> a > 10. I am wondering If that would be a good use for z3.

vlsi commented:
I guess we want to keep RexImplicationChecker, so we don't want to use slow approaches like z3 there.
On the other hand, z3 might be helpful for testing RexImplicationChecker.

qizhou92 commented:
Thank you ! It might be a stupid question. but what is the time requirement in general for checking RexNode implication, like ~10 milliseconds? ~1 milliseconds? ~0.1 milliseconds? Just want to get a sense.

vlsi commented:
>but what is the time requirement in general for checking RexNode implication, like ~10 milliseconds? ~1 milliseconds? ~0.1 milliseconds? Just want to get a sense.

The expectation is that both RexSimplify and RexImplicationChecker have linear complexity with respect to the input subtree size (e.g. the number of nodes).
